sharpd/sharp_nht.c

```diff
@@ -65,3 +69,157 @@ void sharp_nh_tracker_dump(struct vty *vty)
 			nht->updates);
 	}
 }
+
+PREDECL_RBTREE_UNIQ(sharp_nhg_rb);
+
+struct sharp_nhg {
+	struct sharp_nhg_rb_item mylistitem;
+
+	uint32_t id;
+
+	char name[256];
+
+	bool installed;
+};
+
+static uint32_t nhg_id;
+
+static uint32_t sharp_get_next_nhid(void)
+{
+	zlog_debug("Id assigned: %u", nhg_id);
+	return nhg_id++;
+}
+
+struct sharp_nhg_rb_head nhg_head;
+
+static int sharp_nhg_compare_func(const struct sharp_nhg *a,
+				  const struct sharp_nhg *b)
+{
+	return strncmp(a->name, b->name, strlen(a->name));
+}
+
+DECLARE_RBTREE_UNIQ(sharp_nhg_rb, struct sharp_nhg, mylistitem,
+		    sharp_nhg_compare_func);
+
+static struct sharp_nhg *sharp_nhgroup_find_id(uint32_t id)
+{
+	struct sharp_nhg *lookup;
+
+	/* Yea its just a for loop, I don't want add complexity
+	 * to sharpd with another RB tree for just IDs
+	 */
+
+	frr_each (sharp_nhg_rb, &nhg_head, lookup) {
+		if (lookup->id == id)
+			return lookup;
+	}
+
+	return NULL;
+}
+
+static void sharp_nhgroup_add_cb(const char *name)
+{
+	struct sharp_nhg *snhg;
+
+	snhg = XCALLOC(MTYPE_NHG, sizeof(*snhg));
+	snhg->id = sharp_get_next_nhid();
+	strlcpy(snhg->name, name, sizeof(snhg->name));
+
+	sharp_nhg_rb_add(&nhg_head, snhg);
+}
+
+static void sharp_nhgroup_add_nexthop_cb(const struct nexthop_group_cmd *nhgc,
+					 const struct nexthop *nhop)
+{
+	struct sharp_nhg lookup;
+	struct sharp_nhg *snhg;
+	struct nexthop_group_cmd *bnhgc = NULL;
+
+	strlcpy(lookup.name, nhgc->name, sizeof(lookup.name));
+	snhg = sharp_nhg_rb_find(&nhg_head, &lookup);
+
+	if (nhgc->backup_list_name[0])
+		bnhgc = nhgc_find(nhgc->backup_list_name);
+
+	nhg_add(snhg->id, &nhgc->nhg, (bnhgc ? &bnhgc->nhg : NULL));
```

Comment:
this pattern causes the whole new nhg zapi stack to be run once for each nexthop, rather than each nhg - that seems inefficient?

Reply:
Yea you want it for each nexthop so it sends a replace if someone removes/adds a nexthop. Can't do it just on group creation.

Reply:
but - that just seems ... wrong? that would mean MULTIPATH_NUM round trips for full-ecmp groups - complete round trips through zebra and the kernel. can't we identify the point where a configuration change to an nhg occurs - exiting the submode, for example - and that would support both add and modify operations?

Reply:
Sure that makes sense to me. But these callbacks were already in place in lib/nexthop_group.h to subscribe to before this PR. It's not in scope to change them for this imo.